### PR TITLE
Feature/1111 tanf section 4

### LINKED ISSUE #1111: TANF Stratum Data (04) File Parsing - As a data analyst, I want my data files to be parsed so that my submission can be validated/accepted by TDP
Description:
Parsing of TANF STT submissions for Stratum Data.

AC: 
- [ ] Data File parsing logic is in place such that data file submission can be completed by user and parsed by system for full submission once errors are resolved.
- [ ] Data Errors tell user specifically where in the file they occur, and therefore what exactly needs to be fixed in order for the submission to be complete
- [ ] TDP gives the user all data errors for a submission in a way they can easily save or export so they can take the issues back to their case management software and fix the issues then recreate an export for resubmission to TDP.
- [ ] Data Errors must be issued in human readable plain language so that an english speaker can easily understand what is needed. 

Tasks: 
- [ ] Sample tasks

DoD: 
 - [ ] minimum 90% code coverage test
 - [ ] 0 linting errors and warnings
 - [ ] 0 accessibility errors
 - [ ] deployed to a dev environment
 - [ ] all significant changes have been documented
 - [ ] free of high and medium security vulnerabilities

Supporting Documentation:
https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_data_report_section1_10_2008.pdf 

TANF Data Report - Section 4 (Effective 10/1/2008) - Number of Families by Stratum for States that Report Data Based on a Stratified Sample
https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_data_report_section4.pdf 

Open Questions:

(any clarifying questions in support of the issue)

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #1111

Adds parsing/validation for TANF section 4 submissions (T7 data row)

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-backend && docker-compose up
cd tdrs-frontend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Upload the file `ADS.E2J.FTP4.TS06` included in the test data folder to TANF Section 4
3. Verify the `T7` and `ParserError` records are created as expected

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Data File parsing logic is in place such that data file submission can be completed by user and parsed by system for full submission once errors are resolved.
- [x] Data Errors tell user specifically where in the file they occur, and therefore what exactly needs to be fixed in order for the submission to be complete
- [x] TDP gives the user all data errors for a submission in a way they can easily save or export so they can take the issues back to their case management software and fix the issues then recreate an export for resubmission to TDP.
- [x] Data Errors must be issued in human readable plain language so that an english speaker can easily understand what is needed. 
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py]:
```suggestion
                       endIndex=37, required=True, validators=[validators.validateSSN()],
```

Did we miss converting to this new function you made?

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py]: Same here with `validateSSN`

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py]: Hmm, can we improve this `isInLimits(0, 999999)` call? I see we have various fields of different lengths too. I feel like we just need to ensure it's numeric and of a certain length. Do we already have a validator like that? Maybe just two validators instead of this sort of awkward one? The "magic number" 999999* is a code-smell.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py]: Meta question since I'm doing cat4 stuff now -- are postparsing validators always cat3 validators and vice versa?

File: tdrs-backend/tdpservice/parsers/test/factories.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/test/factories.py]: Develop has these t4/t5 factories -- why is this showing up in the PR? I'm confused.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t3.py]: Just hadnt merged develop yet.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py]: As it stands all cat3 validators are post parsing validators and vice versa.

File: tdrs-backend/tdpservice/parsers/test/factories.py
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/test/factories.py]: Should only show T7 now.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: Lets see if we can condense this file under 100 lines, maybe more like 50.

**Review by andrew-jameson — CHANGES_REQUESTED** [review]

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: Nit: Suggestion is to change the constants to capital

File: tdrs-backend/tdpservice/search_indexes/migrations/0017_auto_20230804_1935.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/migrations/0017_auto_20230804_1935.py]: Should wait for the migration ticket to be merged first and then pull those changes in.

File: tdrs-backend/tdpservice/parsers/validators.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/validators.py]: What is zfill doing here?

File: tdrs-backend/tdpservice/search_indexes/migrations/0018_auto_20230810_1153.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/migrations/0018_auto_20230810_1153.py]: WE will need to recreate the migration after migration PR #2703 is merged.

File: tdrs-backend/tdpservice/search_indexes/migrations/0018_auto_20230816_1917.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/migrations/0018_auto_20230816_1917.py:1]: Migration needs to be recreated after merging #2703

File: tdrs-backend/tdpservice/search_indexes/migrations/0017_auto_20230804_1935.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/migrations/0017_auto_20230804_1935.py]: Can do!

**Review by andrew-jameson — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Comment by ADPennington** [comment]
@jtimpe can we get this branch updated and deployed to a dev environment?

**Comment by jtimpe** [comment]
> @jtimpe can we get this branch updated and deployed to a dev environment?

@ADPennington 1111 is now deployed to sandbox

File: tdrs-backend/tdpservice/parsers/test/factories.py
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/test/factories.py]:
```suggestion
    """Generate TANF T7 record for testing."""
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: @jtimpe which data element does 3A map to?

I'm seeing unexpected results. in the screenshots below, 1, 2 and 3 represent months 1, 2, and 3. 

_this is how i expected the data to look after restructuring to one record per tdrs indicator and stratum_
1111p1

_this is how i expected the data to look after restructuring to one record per tdrs indicator, stratum and month_
1111p2

_this is what im seeing in DAC_
1111p3

File: tdrs-backend/tdpservice/search_indexes/admin/tanf.py
Comment (ADPennington) [inline tdrs-backend/tdpservice/search_indexes/admin/tanf.py:112]: i think here we need basically all the data. it is pretty challenging to check the transformation currently. need stratum, tdrs indicator, and families.

 and probably the ability to filter based on TANF vs SSP. and STT but could be a future ticket.

Comment (jtimpe) [reply tdrs-backend/tdpservice/search_indexes/admin/tanf.py:112]: I added additional fields to the `list_display`, but i agree the filtering should be looked at in another ticket if possible. let me know if anything else would be helpful

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: good catch. we were creating schemas starting with index `1` but assumed the first record would have index `0` for the month transform. offsetting the transform value appears to fix it, and i've updated the test cases to reflect that.

File: tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: @jtimpe this is a bit of a nit, but i wonder if users will be confused by the 6A references. in the feedback report. 6A comes up for 6A, 6B, 6C family counts in the feedback report. what do you think @reitermb ? if its an issue, we can wrap it up in #2714 

1111p6
1111p7

Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: this made sense to me so i went ahead and made the change

Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: thanks! i need to make a note of this for section 3 #2713

Comment (ADPennington) [reply tdrs-backend/tdpservice/parsers/schema_defs/tanf/t7.py]: 1111p8

**Review by ADPennington — APPROVED** [review]
excellent work @jtimpe 🚀

**Pull request merged**
